lib/charms/grafana_k8s/v1/grafana_source.py

```diff
@@ -0,0 +1,336 @@
+
+    def snapshot(self) -> Dict:
+        """Save grafana source information"""
+        return {"data": self.data}
+
+    def restore(self, snapshot) -> None:
+        """Restore grafana source information"""
+        self.data = snapshot["data"]
+
+
+class GrafanaSourceEvents(CharmEvents):
+    """Events raised by :class:`GrafanaSourceEvents`"""
+
+    # We are emitting multiple events for the same thing due to the way Grafana provisions
+    # datasources. There is no "convenient" way to tell Grafana to remove them outside of
+    # setting a separate "deleteDatasources" key in the configuration file to tell Grafana
+    # to forget about them, and the reasons why sources_to_delete -> deleteDatasources
+    # would be emitted is intrinsically linked to the sources themselves
+    sources_changed = EventSource(GrafanaSourcesChanged)
+    sources_to_delete_changed = EventSource(GrafanaSourcesChanged)
+
+
+class GrafanaSourceConsumer(ConsumerBase):
+    _stored = StoredState()
+
+    def __init__(
+        self, charm: CharmBase, name: str, consumes: dict, multi=False
+    ) -> None:
+        """Construct a Grafana charm client.
+
+        The :class:`GrafanaSourceConsumer` object provides an interface
+        to Grafana. This interface supports providing additional
+        sources for Grafana to monitor. For example, if a charm
+        exposes some metrics which are consumable by an ingestor
+        (such as Prometheus), then an additional source can be added
+        by instantiating a :class:`GrafanaSourceConsumer` object and
+        adding its datasources as follows:
+
+            self.grafana = GrafanaSourceConsumer(
+                self, "grafana-datasource", {"grafana_datasource"}: ">=2.0"}
+            )
+            self.grafana.add_source(
+                address=<address>,
+                port=<port>
+            )
+
+        Args:
+
+            charm: a :class:`CharmBase` object which manages this
+                :class:`GrafanaSourceConsumer` object. Generally this is
+                `self` in the instantiating class.
+            name: a :string: name of the relation between `charm`
+                the Grafana charmed service.
+            consumes: a :dict: of acceptable monitoring service
+                providers. The keys of the dictionary are :string:
+                names of grafana source service providers. Typically,
+                this is `grafana-source`. The values of the dictionary
+                are corresponding minimal acceptable semantic versions
+                for the service.
+            multi: an optional (default `False`) flag to indicate if
+                this object should support interacting with multiple
+                service providers.
+
+        """
+        super().__init__(charm, name, consumes, multi)
+
+        self.charm = charm
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
```

Comment:
set_default takes `**kwargs`

Reply:
Sure, but I'm not sure it offers any clarity/advantages for two variables.